src/Workspaces/Remote/Core/RemoteCallback.cs

```diff
@@ -87,23 +88,106 @@ public async ValueTask<TResult> InvokeAsync<TResult>(Func<T, CancellationToken,
+                // is passed, which will signal the completion of the writer we have here.  Note: we *do* need to
+                // complete this writer in the event if an exception as that may have happened *prior* to even issuing
+                // the rpc.  If we don't complete the writer we will hang.  If the exception happened within the RPC
+                // the writer may already be completed, but it's fine for us to complete it a second time.
+                //
+                // It is *not* fine for us to complete the writer in a non-exception event as it is legal (and is the
+                // case) that the code in StreamJsonRPC may still be using it (see
+                // https://github.com/AArnott/Nerdbank.Streams/blob/dafeb5846702bc29e261c9ddf60f42feae01654c/src/Nerdbank.Streams/PipeExtensions.cs#L428)
+                // where the writer may be advanced in an independent Task even once the rpc message has returned to the
+                // caller (us). 
+                //
+                // NOTE: it is intentinonal that the try/catch pattern here does NOT match the one in ReadAsync.  There
+                // are very different semantics around each.  The writer code passes ownership to StreamJsonRPC, while
+                // the reader code does not.  As such, the reader code is responsible for completing the reader in all
+                // cases, whereas the writer code only completes when faulting.
+                finally
+                {
+                }
+#endif
+            }
+
+            async Task<TResult> ReadAsync(PipeReader pipeReader)
+            {
+                // NOTE: it is intentinonal that the try/catch pattern here does NOT match the one in WriteAsync.  There
+                // are very different semantics around each.  The writer code passes ownership to StreamJsonRPC, while
+                // the reader code does not.  As such, the reader code is responsible for completing the reader in all
+                // cases, whereas the writer code only completes when faulting.
+
+                Exception? exception = null;
+                try
+                {
+                    return await reader(pipeReader, cancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception ex) when ((exception = ex) == null)
+                {
+                    throw ExceptionUtilities.Unreachable();
+                }
+                finally
+                {
+                    // ensure we always complete the reader so the pipe can clean up all its resources. in the case of
+                    // an exception, attempt to complete the reader with that as well as that will tear down the writer
+                    // allowing it to stop writing and allowing the pipe to be cleaned up.
+                    await pipeReader.CompleteAsync(exception).ConfigureAwait(false);
+                }
```

Comment:
Why this pattern of using an exception filter? The pattern we typically use here is:
```suggestion
                try
                {
                    return await reader(pipeReader, cancellationToken).ConfigureAwait(false);
                }
                catch (Exception ex)
                {
                    // ensure we always complete the reader so the pipe can clean up all its resources. in the case of
                    // an exception, attempt to complete the reader with that as well as that will tear down the writer
                    // allowing it to stop writing and allowing the pipe to be cleaned up.
                    await pipeReader.CompleteAsync(ex).ConfigureAwait(false);
                    throw;
                }
```

Reply:
as per tomas in the last PR this loses valuable informationa bout the stack.